src/storage/page_buffer.c

```diff
@@ -8244,6 +8226,53 @@ pgbuf_read_page_from_file_or_page_server (THREAD_ENTRY * thread_p, const VPID *
 #endif
 	      assert (local_lsa_is_less_than_or_equal_to_page_server_lsa && equal_vpid);
 	    }
+	  else
+	    {
+	      const bool fileio_pages_equal = (io_page->prv == second_io_page->prv);
+	      if (!fileio_pages_equal)
+		{
+		  /* on a transaction server, btree statistics is not written immediately
+		   * to the btree root page; the update happens, for efficiency, only at checkpoint
+		   * time using a in-memory caching mechanism; on a page server, the statistics are
+		   * applied immediately and thus, the difference; compare everything ignoring the
+		   * statistics
+		   */
+		  const PAGE_PTR pgptr = (PAGE_PTR) ((char *) io_page->page);
+		  PAGE_PTR second_pgptr = (PAGE_PTR) ((char *) second_io_page->page);
+		  if (io_page->prv.ptype == PAGE_BTREE && second_io_page->prv.ptype == PAGE_BTREE
+		      && btree_is_btree_root_page (thread_p, pgptr)
+		      && btree_is_btree_root_page (thread_p, second_pgptr))
+		    {
+		      // page server might already contain btree stats which are still kept in memory on the
+		      // active transaction server
+		      assert (io_page->prv.lsa <= second_io_page->prv.lsa
+			      && io_page->prv.pageid == second_io_page->prv.pageid
+			      && io_page->prv.volid == second_io_page->prv.volid
+			      && io_page->prv.pflag == second_io_page->prv.pflag
+			      && io_page->prv.pflag == second_io_page->prv.pflag
+			      && io_page->prv.tde_nonce == second_io_page->prv.tde_nonce);
+		    }
+		  else
+		    {
+#if !defined(NDEBUG)
+		      if (!fileio_pages_equal)
+			{
+			  er_log_debug (ARG_FILE_LINE, "pgbuf_read_page_from_file_or_page_server"
+					" past recovery redo pages not equal\n"
+					"    target repl LSA: %lld|%d\n"
+					"    local page  VPID: %d|%d  LSA: %lld|%d  ptype: %d\n"
+					"    remote page VPID: %d|%d  LSA: %lld|%d  ptype: %d\n",
+					LSA_AS_ARGS (&target_repl_lsa),
+					io_page->prv.volid, io_page->prv.pageid,
+					LSA_AS_ARGS (&io_page->prv.lsa), io_page->prv.ptype,
+					second_io_page->prv.volid, second_io_page->prv.pageid,
+					LSA_AS_ARGS (&second_io_page->prv.lsa), second_io_page->prv.ptype);
+			}
+#endif
+		      assert (fileio_pages_equal);
```

Comment:
I guess that assert must be removed.

Reply:
The assert must stay. If the page is not a btree root page (checked in the previous if), then check must fail.

Reply:
My expectation was assert (false), but is ok.